### PR TITLE
Beefed out the documentation for DataMapper, fields, filters and Mode…

### PR DESCRIPTION
**Pull request by limbera**
Beefed out the documentation for DataMapper, fields, filters and ModelMapper

### CODE REVIEW

File: docs/datamapper/index.rst [hunk 1]
```diff
@@ -1,51 +1,205 @@
-============
-Data Mappers
-============
+===========
+DataMappers
+===========
 
-DataMappers, as the name suggest, map properties on themselves to your object.
+DataMappers function to reduce Python objects into simple data types supported
```
Comment (funkybob) [inline docs/datamapper/index.rst]: function _is_ to reduce...

File: docs/datamapper/index.rst [hunk 2]
```diff
@@ -1,51 +1,205 @@
...
-DataMappers, as the name suggest, map properties on themselves to your object.
+DataMappers function to reduce Python objects into simple data types supported 
+by JSON, and to expand JSON into Python objects. They are an alternative 
+approach to using Serialisers. 
 
-They allow you to easily write proxy objects, primarily for converting between
-serialised and live formats of your data.
+As the name suggests, a DataMapper will map properties on themselves to your 
+object. They allow you to easily write proxy objects, primarily for converting 
+between serialised (JSON) and live (Python) formats of your resources.
```
Comment (funkybob) [inline docs/datamapper/index.rst]: Isn't this just repeating the last paragraph?

Comment (limbera) [inline docs/datamapper/index.rst]: hmm seems so! pick which one you prefer, or amalgamate them?

File: docs/datamapper/index.rst [hunk 3]
```diff
@@ -1,51 +1,205 @@
...
+    to, even when using << syntax, instances MUST NOT be shared between threads.
 
-When constructing a DataMapper you can pass an object for it to "bind" to. All
-field access to the DataMapper fields will proxy to this bound object.
+Field decorator: get/set
+========================
 
-Thread Safety
-=============
+You can set and get properties on a DataMapper by using Python's descriptor
```
Comment (funkybob) [inline docs/datamapper/index.rst]: reword...

    DataMappers leverage Python's "descriptor protocol" to allow you to control how properties are read from and written to your objects.

File: docs/datamapper/index.rst [hunk 4]
```diff
@@ -1,51 +1,205 @@
...
+            self.is_alive = is_alive
+
+
+    from nap import datamapper
+
+    # A DataMapper that we are creating for the Person object
+    class UserMapper(datamapper.DataMapper):
+
+        '''
+        The self argument refers to the object we bind to the DataMapper when
```
Comment (funkybob) [inline docs/datamapper/index.rst]: "object we bind the DataMapper to"

File: docs/datamapper/index.rst [hunk 5]
```diff
@@ -1,51 +1,205 @@
...
+    # A DataMapper that we are creating for the Person object
+    class UserMapper(datamapper.DataMapper):
+
+        '''
+        The self argument refers to the object we bind to the DataMapper when 
+        we construct it. It DOES NOT refer to the instance of the UserMapper.
+        '''
+        @datamapper.field 
+        def name(self):
+            return {}.format(self.first_name, self.last_name)
```
Comment (funkybob) [inline docs/datamapper/index.rst]: Is this GH fucking with the formatting?  Shouldn't that be "{} {}" ?

File: docs/datamapper/index.rst [hunk 6]
```diff
@@ -1,51 +1,205 @@
...
+
+        # We can use the Field class for simpler cases
+        first_name = datamapper.Field('first_name')
+        last_name = datamapper.Field('last_name')
+        is_alive = datamapper.Field('is_alive')
+
+    # Construct instances of the Person and a DataMapper clases
+    person = Person('Jane', 'Doe', 22, True)
+    mapper = DataMapper(person)
```
Comment (funkybob) [inline docs/datamapper/index.rst]: Perhaps follow this with some samples of using the DM, yielding results...

File: docs/datamapper/index.rst [hunk 7]
```diff
@@ -1,51 +1,205 @@
...
+    class UserMapper(datamapper.DataMapper):
+
+        @datamapper.field 
+        def name(self):
+            return {}.format(self.first_name, self.last_name)
+
+        first_name = datamapper.Field('first_name')
+        last_name = datamapper.Field('last_name')
+        is_alive = datamapper.Field('is_alive', filters=[BooleanFilter])
+
```
Comment (funkybob) [inline docs/datamapper/index.rst]: Mention that when they fail, a ValidationError is raised.

File: docs/datamapper/index.rst [hunk 8]
```diff
@@ -1,51 +1,205 @@
...
+ModelDataMappers
+================
+A ModelDataMapper will automatically create a DataMapper for a Django model.
```
Comment (funkybob) [inline docs/datamapper/index.rst]: Liken this to a ModelForm...

File: docs/datamapper/index.rst [hunk 9]
```diff
@@ -1,51 +1,205 @@
...
+
+
+A ModelDataMapper behaves very similar to a Django ModelForm, you use it by 
+setting some fields in an inner Meta class. The fields that can be set are:
+
+- model (default = None)
+- fields (default = [])
+- exclude (default = [])
+- required (default = {})
+- queryset (default = None)
```
Comment (funkybob) [inline docs/datamapper/index.rst]: This does not exist...

File: docs/datamapper/index.rst [hunk 10]
```diff
@@ -1,51 +1,205 @@
...
+        class Meta:
+            model = models.Person
+            fields = '__all__'
 
-   class UserMapper(DataMapper):
-       @field
-       def name(self):
-           return self.get_full_name()
-       name.required = False
+You can still use the `property` built-in to get/set properties on a
```
Comment (funkybob) [inline docs/datamapper/index.rst]: and Field

File: docs/datamapper/index.rst [hunk 11]
```diff
@@ -1,51 +1,205 @@
...
+            fields = '__all__'
 
-   class UserMapper(DataMapper):
-       @field
-       def name(self):
-           return self.get_full_name()
-       name.required = False
+You can still use the `property` built-in to get/set properties on a 
+ModelDataMapper. This is useful when the model contains some properties that
+the ModelDataMapper cannot understand.
```
Comment (funkybob) [inline docs/datamapper/index.rst]: or when you want to customise how certain fields are represented.

File: docs/datamapper/index.rst [hunk 12]
```diff
@@ -1,51 +1,276 @@
...
+    to, even when using << syntax, instances MUST NOT be shared between threads.
 
-They allow you to easily write proxy objects, primarily for converting between
-serialised and live formats of your data.
 
-DataMappers work using Python's descriptor protocol, which is most commonly
-used via the `property` built-in.
+Field decorator: get/set
+========================
+You can set and get properties on a DataMapper using Python's descriptor
```
Comment (funkybob) [inline docs/datamapper/index.rst]: I still think this needs rewording.

**Pull request merged**